protocol/0082-RFPR-on_chain_referral_program.md

```diff
@@ -0,0 +1,235 @@
+
+## Benefit mechanics
+
+### Setting benefit factors
+
+Whilst a referral program is `STATUS_ACTIVE`, at the start of an epoch the network must set the `referral_reward_factor` and `referral_discount_factor` for each referee. This is done by identifying a referees team and identifying the teams current benefit tier. A teams benefit tier is defined as the highest tier for which their `team_running_volume` is greater or equal to the tiers `minimum_running_volume`. If a party does not qualify for any tier, both values are set to `0`.
+
+```psuedo
+Given:
+    benefit_tiers=[
+        {
+            "minimum_running_volume": 10000,
+            "referral_reward_factor": 0.001,
+            "referral_discount_factor": 0.001,
+        },
+        {
+            "minimum_running_volume": 20000,
+            "referral_reward_factor": 0.005,
+            "referral_discount_factor": 0.005,
+        },
+        {
+            "minimum_running_volume": 30000,
+            "referral_reward_factor": 0.010,
+            "referral_discount_factor": 0.010,
+        },
+    ]
+
+And:
+    team_running_volume=22353
+
+Then:
+    referral_reward_factor=0.005
+    referral_discount_factor=0.005
+```
+
+These benefit factors are then fixed for the duration of the next epoch.
+
+### Applying benefit factors
+
+Whenever a fee is incurred by a referee (either during continuous trading or on auction exit) the network must apply referral program benefits.
+
+The network can first calculate individual fee components following the [fees specification](./0029-FEES-fees.md#calculating-fees) and apply any usual checks.
+
+The network must then:
+
+- Calculate the rewards due to the referrer.
+    ```pseudo
+    infrastructure_fee_reward = floor(infrastructure_fee * referral_reward_factor) 
+    liquidity_fee_reward = floor(liquidity_fee * referral_reward_factor) 
+    maker_fee_reward = floor(maker_fee * referral_reward_factor) 
+    ```
+- Calculate the discounts due to the referee.
+    ```pseudo
+    infrastructure_fee_discount = floor(infrastructure_fee * referral_discount_factor)
+    liquidity_fee_discount = floor(liquidity_fee * referral_discount_factor)
+    maker_fee_discount = floor(maker_fee * referral_discount_factor)
+    ```
+- And then update the fees.
+    ```pseudo
+    infrastructure_fee = infrastructure_fee - infrastructure_fee_reward - infrastructure_fee_discount
+    liquidity_fee = liquidity_fee - liquidity_fee_reward - liquidity_fee_discount
+    maker_fee = maker_fee - maker_fee_reward - maker_fee_discount
+    ```
+
+Note the rewards and discounts are floored rather than raised to ensure the final infrastructure fee cannot be negative.
+
+The network can then carry out the normal fee transfers using the updated fee amounts followed by additional transfers from the referees general account to the referrers general account. These transfers will use the following new transfer types.
```

Comment:
Should there instead be a new fee type which goes directly to the referrer rather than just a transfer? And maybe should be clear here that it is the `infrastructure_fee_reward` listed above (and maybe rename that to `_referrer_reward` or something. It threw me at first to see that as a discount to the fee the referee was paying.